File: Assets/Scripts/Domain/Defs/CombinationDef.cs
using System.Linq;
using UnmappedIsland.Domain.Runtime;

namespace UnmappedIsland.Domain.Defs
{
    /// <summary>
    /// ドラッグ型のカード間相互作用（GameElementDefinition.md 12節）。ドロップされた側（受け側）の
    /// object_defに定義する。Withは、ドラッグされてきたカードとのマッチング条件（タグのグローバルID、12.1節）。
    /// conditionsと実行結果（EffectOutcome＝active/pick）を1つの定義としてまとめて持つ。
    /// </summary>
    public sealed class CombinationDef
    {
        public string Name { get; }
        private readonly int with;

        /// <summary>nullなら常に真（conditions省略）。</summary>
        private readonly ConditionNode conditions;

        /// <summary>条件成立時の実行結果（active/pickの解決はEffectOutcome自身が行う）。</summary>
        private readonly EffectOutcome outcome;

        public CombinationDef(
            string name,
            int with,
            ConditionNode conditions,
            EffectOutcome outcome)
        {
            Name = name;
            this.with = with;
            this.conditions = conditions;
            this.outcome = outcome;
        }

        /// <summary>draggedDefがこのcombinationのWithタグを持っていれば真（12.1節）。</summary>
        public bool Matches(ObjectDef draggedDef) => draggedDef.Tags.Contains(with);

        public bool TryExecute(WorldObject self, WorldObject dragged, WorldObject actor, WorldSession session)
        {
            if (!Matches(dragged.Def)) return false;
            if (conditions != null && !conditions.Evaluate(root => ReferenceRootResolver.Resolve(root, self, actor, dragged)))
                return false;

            ActiveEffect effect = outcome.Resolve(self, actor, dragged, session);
            if (effect != null) self.ApplyActiveEffect(effect, session, actor, dragged);
            return true;
        }
    }
}

File: src/domain/defs/CombinationDef.ts
import type { WorldObject } from '../runtime/WorldObject';
import type { WorldSession } from '../runtime/WorldSession';
import type { ActiveEffect } from './ActiveEffect';
import type { CraftingInput } from './CraftingStep';
import type { DefNames, DescriptionWriter } from './Description';
import { tagRef, text } from './Description';
import { InteractionDef } from './InteractionDef';
import type { ObjectDef } from './ObjectDef';
import type { WeightSpec } from './PickEffect';
import type { Requirements } from './Requirement';

/**
 * ドラッグ型のカード間相互作用（GameElementDefinition.md 12節）。素材側のobject_defに1つだけ定義し、
 * 宣言している側がself・相手がdraggedになる（どちらの札を掴んでも同じ宣言が実行される、12.3節）。
 * withは、相手とのマッチング条件（タグのグローバルID、12.1節）。
 */
export class CombinationDef extends InteractionDef {
  private readonly with: number;

  constructor(
    name: string,
    withTagGlobalId: number,
    requirements: Requirements | undefined,
    effect: ActiveEffect,
    duration?: WeightSpec,
  ) {
    super(name, requirements, effect, duration);
    this.with = withTagGlobalId;
  }

  protected describeTrigger(names: DefNames, out: DescriptionWriter): void {
    out.write(text('with: '), tagRef(names.tagName(this.with)), text('を持つカードのドロップ'));
  }

  protected get craftingKind(): 'combination' {
    return 'combination';
  }

  /** ドラッグされてくる相手はタグで指される。消費されるかはdraggedへのdestroyの有無から分かる。 */
  protected override extraCraftingInputs(effect: ActiveEffect): readonly CraftingInput[] {
    return [{ kind: 'tag', tagGlobalId: this.with, consumed: effect.destroys('dragged') }];
  }

  /** draggedDefがこのcombinationのwithタグを持っていれば真（12.1節）。 */
  matches(draggedDef: ObjectDef): boolean {
    return draggedDef.tags.includes(this.with);
  }

  tryExecute(
    self: WorldObject,
    dragged: WorldObject,
    actor: WorldObject | undefined,
    session: WorldSession,
  ): boolean {
    return this.matches(dragged.def) && this.apply(self, dragged, actor, session);
  }
}
